xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-notifiers/xwiki-platform-notifications-notifiers-default/src/main/java/org/xwiki/notifications/notifiers/internal/UserEventManager.java

```diff
@@ -302,4 +315,30 @@ private boolean isEventFiltered(List<NotificationFilter> filters, Event event, D
         }
         return false;
     }
+
+    public void cleanUpFilters(DocumentDeletedEvent event, DocumentReference user)
+    {
+        DocumentReference deletedDocumentReference = event.getDocumentReference();
+        String serializedReference = this.entityReferenceSerializer.serialize(deletedDocumentReference);
+        try {
+            Set<NotificationFilterPreference> matchingPreferences =
+                this.notificationFilterPreferenceManager.getFilterPreferences(user)
+                    // We only clean up the filters matching exactly the page reference
+                    // we never clean up filters for space / wiki there
+                    .stream().filter(pref -> StringUtils.equals(pref.getPageOnly(), serializedReference))
+                    .collect(Collectors.toSet());
+            if (!matchingPreferences.isEmpty()) {
+                CleaningFilterEvent cleaningFilterEvent = new CleaningFilterEvent();
+                this.observationManager.notify(cleaningFilterEvent, deletedDocumentReference, matchingPreferences);
+                if (!cleaningFilterEvent.isCanceled()) {
+                    for (NotificationFilterPreference matchingPreference : matchingPreferences) {
+                        this.notificationFilterPreferenceManager.deleteFilterPreference(user,
```

Comment:
Would that be ok or do we need an API for performing the delete in batch? I have the feeling it should be ok: in theory we shouldn't get that many filters preferences to delete...

Reply:
I actually provided the API to delete in batch.